### PR TITLE
arch: X86: update x86 arch porting with new acpica interface

### PR DESCRIPTION
**Pull request by najumon1980**
update x86 arch porting with new acpica interface. As part of this the old x86 acpi implementation has been removed and updated test apps and other board configurations with new  acpica interface support.

### CODE REVIEW

**Comment by najumon1980** [comment]
Will add board/soc update for other x86 platforms once tested

**Comment by najumon1980** [comment]
Few CI job are failed and this will resolve once acpica PR get merged.

**Comment by najumon1980** [comment]
@andyross could you please have a look into the changes ?

File: include/zephyr/acpi/acpi.h [hunk 1]
```diff
@@ -117,9 +126,30 @@ int acpi_device_type_get(ACPI_RESOURCE *res);
...
+ * @brief retrieve acpi MAD table for the given type.
+ *
+ * @param type type of requested MAD table
+ * @param tables pointer to the MAD table
+ * @param num_inst number of instance for the requested table
+ * @param max_inst maximum number of MAD table buffer
  * @return return 0 on success or error code
  */
-int acpi_table_get(char *signature, int inst, void **acpi_table);
+int acpi_get_madt(int type, struct acpi_subtable_header **tables, int *num_inst, int max_inst);
```
Comment (tbursztyka) [inline include/zephyr/acpi/acpi.h:142]: acpi sub-tables are all contiguous and persistent. Why one would need to allocate another place-holder, with the risk of over-allocating (memory loss) and under-allocating (information loss).

Just follow what has been done so far.

File: lib/acpi/Kconfig [hunk 2]
```diff
@@ -52,3 +52,9 @@ config ACPI_HID_LEN_MAX
 	default 12
 	help
 	  Size of HID string.
+
+config ACPI_MADT_LAPIC_MAX
```
Comment (tbursztyka) [inline lib/acpi/Kconfig:62]: So this type of options would be totally useless.

File: lib/acpi/acpi.c [hunk 3]
```diff
@@ -224,8 +223,8 @@ static ACPI_STATUS acpi_enable_pic_mode(void)
 	return status;
 }
 
-static int acpi_get_irq_table(struct acpi *bus, char *bus_name,
-				ACPI_PCI_ROUTING_TABLE *rt_table, uint32_t rt_size)
+static int acpi_get_irq_table(struct acpi *bus, char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table,
```
Comment (tbursztyka) [inline lib/acpi/acpi.c:226]: why changing this?

File: lib/acpi/acpi.c [hunk 4]
```diff
@@ -491,8 +492,7 @@ int acpi_current_resource_free(ACPI_RESOURCE *res)
 	return 0;
 }
 
-int acpi_get_irq_routing_table(char *bus_name,
-			       ACPI_PCI_ROUTING_TABLE *rt_table, size_t rt_size)
+int acpi_get_irq_routing_table(char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table, size_t rt_size)
```
Comment (tbursztyka) [inline lib/acpi/acpi.c:495]: same here, why?

File: lib/acpi/acpi.c [hunk 5]
```diff
@@ -633,17 +633,137 @@ int acpi_table_get(char *signature, int inst, void **acpi_table)
 		status = acpi_early_init();
 		if (status) {
 			LOG_ERR("ACPI early int failed");
```
Comment (tbursztyka) [inline lib/acpi/acpi.c:635]: s/int/init

File: lib/acpi/acpi.c [hunk 6]
```diff
@@ -633,17 +633,137 @@ int acpi_table_get(char *signature, int inst, void **acpi_table)
...
 	status = AcpiGetTable(signature, inst, &table);
 	if (ACPI_FAILURE(status)) {
 		LOG_ERR("ACPI get table failed: %d", status);
+		return NULL;
+	}
+
+	return (void *)table;
+}
+
+int acpi_get_madt(int type, struct acpi_subtable_header **tables, int *num_inst, int max_inst)
```
Comment (tbursztyka) [inline lib/acpi/acpi.c:649]: actually the function is not even doing what its name means: it is not returning the madt table, but fills in an array of pointer of all (as long as allocation fits, see previous comment), sub-table pointer. 

"acpi_get_madt_entry" would somehow be more semantically right.

File: lib/acpi/acpi.c [hunk 7]
```diff
@@ -633,17 +633,137 @@ int acpi_table_get(char *signature, int inst, void **acpi_table)
...
+	}
+
+	*num_inst = i;
+	if (!i) {
+		return -ENODEV;
+	}
+	return 0;
+}
+
+static int acpi_dmar_get(enum AcpiDmarType type, struct acpi_subtable_header **tables,
```
Comment (tbursztyka) [inline lib/acpi/acpi.c:682]: same here, this does not return dmar table (and has the same allocation issue).

File: modules/acpica/Kconfig [hunk 8]
```diff
@@ -6,4 +6,10 @@ config ACPI
 
 menu "ACPI driver support"
 
+config ACPI_FULL_SUPPORT
```
Comment (tbursztyka) [inline modules/acpica/Kconfig:9]: nah, either we support acpi or we don't. Let's not add intermediate support (which is cryptic as well: what means CONFIG_ACPI alone vs CONFIG_ACPI_FULL_SUPPORT then?)

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Instead of porting, you change the logic. Which is premature as well: for instance, we are not going to support more than one ioapic anytime soon (if ever, actually).

plus various unnecessary changes etc..

**Comment by najumon1980** [comment]
@tbursztyka update as per your review comment and also responded to your few comments

**Comment by jhedberg** [comment]
@najumon1980 looks like there's a bunch of build failures:
```
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsReadMemory':
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:390:31: error: implicit declaration of function 'sys_read64'; did you mean 'sys_read32'? [-Werror=implicit-function-declaration]
  390 |         *((UINT64 *) Value) = sys_read64 (Address);
      |                               ^~~~~~~~~~
      |                               sys_read32
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsWriteMemory':
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:441:9: error: implicit declaration of function 'sys_write64'; did you mean 'sys_write32'? [-Werror=implicit-function-declaration]
  441 |         sys_write64 ((UINT64) Value, Address);
      |         ^~~~~~~~~~~
      |         sys_write32
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsGetRootPointer':
/__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:778:13: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
  778 |     return ((ACPI_PHYSICAL_ADDRESS) efi_get_acpi_rsdp ());
      |             ^
```

**Comment by najumon1980** [comment]
> @najumon1980 looks like there's a bunch of build failures:
> 
> ```
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsReadMemory':
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:390:31: error: implicit declaration of function 'sys_read64'; did you mean 'sys_read32'? [-Werror=implicit-function-declaration]
>   390 |         *((UINT64 *) Value) = sys_read64 (Address);
>       |                               ^~~~~~~~~~
>       |                               sys_read32
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsWriteMemory':
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:441:9: error: implicit declaration of function 'sys_write64'; did you mean 'sys_write32'? [-Werror=implicit-function-declaration]
>   441 |         sys_write64 ((UINT64) Value, Address);
>       |         ^~~~~~~~~~~
>       |         sys_write32
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c: In function 'AcpiOsGetRootPointer':
> /__w/zephyr/modules/lib/acpica/source/os_specific/service_layers/oszephyr.c:778:13: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
>   778 |     return ((ACPI_PHYSICAL_ADDRESS) efi_get_acpi_rsdp ());
>       |             ^
> ```

This need following acpica PR need to merge and update the west.yml accordingly. 
https://github.com/acpica/acpica/pull/882

**Comment by jhedberg** [comment]
> This need following acpica PR need to merge and update the west.yml accordingly. [acpica/acpica#882](https://github.com/acpica/acpica/pull/882)

Please submit a PR to our ACPICA mirror (which will not be merged - we’ll sync the tree when the ACPICA upstream PR is merged), and follow up these instructions: https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules

That way we can verify that CI passes with the combination of the ACPICA PR and this PR.

File: modules/acpica/Kconfig (hunk 8 above)
Comment (tbursztyka) [reply modules/acpica/Kconfig:9]: So basically, it's the difference between getting acpica that enables only basic ACPI tables and acpica that enables the same plus DSDT support? 

In that case, name this option ACPI_DSDT_SUPPORT, it's clearer imo.

**Review by tbursztyka — DISMISSED** [review]
Fix the Kconfig option as suggested. 

Rest lgtm

File: include/zephyr/arch/x86/x86_acpi_osal.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ */
+#include <zephyr/arch/x86/efi.h>
+#include <zephyr/arch/x86/legacy_bios.h>
+
+#ifndef ZEPHYR_ARCH_X86_INCLUDE_X86_ACPI_H_
+#define ZEPHYR_ARCH_X86_INCLUDE_X86_ACPI_H_
+
+uint32_t bios_acpi_rsdp_get(void);
+
+#if defined(CONFIG_X86_EFI)
```
Comment (tbursztyka) [inline include/zephyr/arch/x86/x86_acpi_osal.h:14]: No you can't assume that once EFI is enabled, the system will always get it successfully from EFI.

Look at the original code you removed in arch/x86/core/acpic.c: is efi fails to get rsdp, it falls back to legacy lookup.

File: include/zephyr/acpi/acpi.h [hunk 10]
```diff
@@ -19,6 +19,22 @@ struct acpi_dev {
...
+		uint16_t function: 3;
+		uint16_t device: 5;
+		uint16_t bus: 8;
+	} bits;
+
+	uint16_t raw;
+};
+
+struct acpi_mcfg {
+	struct acpi_table_header Header;
```
Comment (tbursztyka) [inline include/zephyr/acpi/acpi.h:33]: why the capital letter ? 

Unless we use an externally declared structures/variable/etc... (so from hal, lib etc...), we do not use came case in Zephyr. So here, just use small letter.

File: tests/kernel/smp/src/main.c [hunk 11]
```diff
@@ -372,7 +372,7 @@ ZTEST(smp, test_coop_resched_threads)
 	 * since we don't give up current CPU, last thread
 	 * will not get scheduled
 	 */
-	spawn_threads(K_PRIO_COOP(10), num_threads, !EQUAL_PRIORITY,
+	spawn_threads(K_PRIO_COOP(MAX_NUM_THREADS), num_threads, !EQUAL_PRIORITY,
```
Comment (tbursztyka) [inline tests/kernel/smp/src/main.c:375]: how does this relate to ACPICA? (as above in the smp's prj.conf)

Make it its own commit, and have a proper commit message explaining the change.

File: tests/arch/x86/info/src/acpi.c [hunk 12]
```diff
@@ -4,145 +4,47 @@
...
 	if (nr_cpus == 0) {
 		printk("ACPI: no RSDT/MADT found\n\n");
 	} else {
 		printk("ACPI: %d CPUs found\n", nr_cpus);
 
-		for (int i = 0; i < nr_cpus; ++i) {
-			struct acpi_cpu *cpu = z_acpi_get_cpu(i);
-			printk("\tCPU #%d: APIC ID 0x%02x\n", i, cpu->apic_id);
+		for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; ++i) {
+			printk("\tCPU #%d: APIC ID 0x%02x\n", i, lapic[i]->Id);
```
Comment (tbursztyka) [inline tests/arch/x86/info/src/acpi.c:47]: No this is bogus: it can happen that a defined CPU in madt might not be enabled at all,  so we don't want to display those. Thus the logic above in acpi() that first counts the valid CPUs. Don't change that.

File: tests/arch/x86/info/src/acpi.c [hunk 13]
```diff
@@ -4,145 +4,47 @@
...
-		path = (struct acpi_dmar_dev_path *)(POINTER_TO_UINT(path) +
-						     ACPI_DMAR_DEV_PATH_SIZE);
-	}
-
-	printk("\n");
+	printk("IOAPIC Info: Bus: %d, dev:%d, fun:%d\n", dmar_id->bits.bus, dmar_id->bits.device,
+	       dmar_id->bits.function);
 }
 
-static void vtd_drhd_info(struct acpi_drhd *drhd)
```
Comment (tbursztyka) [inline tests/arch/x86/info/src/acpi.c:59]: why removing this this? Keep it.

File: tests/arch/x86/info/src/acpi.c [hunk 14]
```diff
@@ -4,145 +4,47 @@
...
 {
-	struct acpi_dmar_dev_path *path;
-	uint16_t id;
-	int n_path;
-
-	printk("\t\t\t. Type: ");
+	union acpi_dmar_id *dmar_id;
+	int num_inst;
 
-	switch (dev_scope->type) {
```
Comment (tbursztyka) [inline tests/arch/x86/info/src/acpi.c:17]: why removing this? Keep it.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Still quite some issues.

First: have proper commit message. Copy/pasting the commit title in the message is not a proper message. Unless the change is supper trivial, you need to give a little explanation on what the commit is doing.

Second: when you port code to a new interface, you'll need to port existing logic and **not** removing it. See my comments in tests/arch  etc...

Also, let's wait now that ACPICA gets the PR 882 applied, let's update the west manifest and then let's see if CI in this present PR passes.

File: tests/arch/x86/info/src/acpi.c (hunk 12 above)
Comment (tbursztyka) [reply tests/arch/x86/info/src/acpi.c:47]: Nevermind this comment, I forgot that z_acpi_get_cpu() does the check for disablement. 

and yeah, the above logic is used only to verify that we got rsdt/madt.

**Review by tbursztyka — DISMISSED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]
